### PR TITLE
Overhaul crypto (Schnorr/Ristretto, HDKD, BIP39)

### PR DESCRIPTION
**Pull request by gavofyork**
- Make Sr25519 the default crypto for accounts/transaction signing.
- Introduces a new "Secret URI" format for specifying secret keys `<phrase>/<soft-key>//<hard-key>///<password>` (the `///password` may be omitted, and `/<soft-key>` and `//<hard-key>` maybe repeated and mixed).
  - Derivation path items may be numeric (encoded as 256-bit LE) or textual (zero-padded to the right unless greater than 32-characters, in which case hashed using blake2).
  - Sr25519 supports BIP39 phrases, a password and a derivation path of any number of soft and hard keys.
  - Ed25519 supports BIP39 phrases, a password and a derivation path of any number of hard keys.
  - Sr25519's Public key type can also be derived now (though nothing uses it yet).
  - SURIs can be used in `subkey` and on the CLI for the `--key` parameter.
- Untangles Authority signing (which remains Ed25519) from Account (extrinsic) signing (which is Sr25519).
  - The two different keyrings and named `AccountKeyring` and `AuthorityKeyring` accordingly.
  - Some dependent hashes in tests have been removed in order to minimise faff when moving between signing schemes.
  - Alice, Bob &c. keys are now formed as SURIs with a single well known seed phrase and different paths for each identity, which is somewhat cleaner than the whitespace padded seeds from before.
- This also revamps the use of the crypto:
  - ed25519/sr25519 `Pair`s are now trait impls.
  - `AuthorityId` and `AccountId` are now type-safe throughout the codebase (`ed25519::Public` or `sr25519::Public`) and not just opaque, weakly typed `H256`s.
  - This removes the need for a lot of crap code when dealing with `Keyring`; a lot less `.to_raw_public`, `.0` and `.into`.
  - Removed all mentions of the `Ed25519AuthorityId` non-type, replacing with local aliasing of `ed25519::Public` to `AuthorityId`, ready for a later refactoring that can replace specific sets of instances with BLS or other crypto.
  - Remove a lot of the highly dodgy coercions between strictly typed public keys and weakly typed arrays (that happen to be the same length).
  - Introduce `UncheckedFrom` and `UncheckedInto` traits for the times (actually, there's only one instance) where we do need to break strong typing and force coercion.

TODO:
- [x] Rejig `Keyring::One` and `::Two`.
- [x] Remove hard-coded hashes from `test-runtime/system.rs` tests.
- [x] Make `subkey` support public key derivation querying.
- [x] Make `subkey` support signing.

CC @jacogr

### CODE REVIEW

**Comment by pepyakin** [comment]
This doesn't rebuild runtime wasm artifacts. Should it?

**Comment by burdges** [comment]
I added an `attach_rng` function in https://github.com/w3f/schnorrkel/commit/9f203515f54ce6e4f9a094251fa9063dfc6acd73#diff-e0934bb6f70a4f8c3baf22a773c14ec7R331 in case the bindings need manual control over the randomness.

**Comment by jacogr** [comment]
@pepyakin yes, probably along with a version bump as well so detection of which version to use is possible for APIs.

Thank you though, was wondering about local tests not yielding the expected results, I needed to rebuild.

**Comment by gguoss** [comment]
Not compatible with consensus and finality-grandpa modules？ Because these modules are hardcoded ed25519.

File: node/cli/src/chain_spec.rs [hunk 1]
```diff
@@ -40,25 +40,28 @@ pub fn dried_danta_config() -> Result<ChainSpec, String> {
...
-		hex!["7e58b096b95c4b3b271f27fedd9f2c51edd48b9d37046240e601180c9dcc8c27"].into(),	// 5EvNEhYYd4b9giczuCo2o8bfLZoKW9jnTeUukfL1NWsAAeEx TODO: change once we switch to sr25519
-		hex!["36dfc933bb0848d8addf16a961369b2e122633a5819a19e43c8142381a1280e3"].into(),	// 5DJevPKpz4EEvmSpK7W6KemS3i5JYPq5FEuEewgRY2cZCxNg TODO: change once we switch to sr25519
+		hex!["e269e835e0bc07c497d55bc17c7bb29c85c5615f9e61582ffdeca7e5f5c66578"].into(), // 5HBa95U5HDFCV1N5Xyrjti65F71tHRQcPbZBmkxRJ39SpqzM
+		hex!["3e9829e6fd4fc7501b504fc16f12177c6c7f38aeb3b8344efb9b15ee85118b2c"].into(), // 5DUn2afs2QevZ6PrGu8snrt76157oacH6JXUD8JNM18VKMwK
 		hex!["8101764f45778d4980dadaceee6e8af2517d3ab91ac9bec9cd1714fa5994081c"].into(),
 	)];
+	// generated with secret: subkey -p danta-root restore $secret
 	let endowed_accounts: Vec<AccountId> = vec![
-		hex!["f295940fa750df68a686fcf4abd4111c8a9c5a5a5a83c4c8639c451a94a7adfd"].into(),	// 5HYmsxGRAmZMjyZYmf7uGPL2YDQGHEt6NjGrfUuxNEgeGBRN TODO: change once we switch to sr25519
+		hex!["343df6f04ffae0840f214f6cb0da00b612c7e9347f980e7afafc520582f79136"].into(), // 5DFCkiP9vky31C1ZP3LpuQYinLAFwQqq6vda7NXa8ALCpq5D TODO: change once we switch to sr25519
```
Comment (tomusdrw) [inline node/cli/src/chain_spec.rs]: Is the TODO still valid?

Comment (gavofyork) [reply node/cli/src/chain_spec.rs]: no...

**Comment by gavofyork** [comment]
@Kianenigma @tomusdrw could do with another review.
@rphmeier you should check the grandpa module - there was an invalid conversion that i removed there.

File: node/executor/src/lib.rs [hunk 2]
```diff
@@ -318,10 +316,10 @@ mod tests {
 			sudo: Some(Default::default()),
 			grandpa: Some(GrandpaConfig {
 				authorities: vec![ // set these so no GRANDPA events fire when session changes
-					(keyring::ed25519::Keyring::Charlie.to_raw_public().into(), 1),
-					(keyring::ed25519::Keyring::Bob.to_raw_public().into(), 1),
-					(keyring::ed25519::Keyring::Alice.to_raw_public().into(), 1),
-				],
+/*					(AuthorityKeyring::Charlie.to_raw_public().into(), 1),
```
Comment (tomusdrw) [inline node/executor/src/lib.rs]: Any reason to keep the commented out code in that file?

File: srml/grandpa/src/lib.rs [hunk 3]
```diff
@@ -340,19 +340,14 @@ impl<T> Default for SyncedAuthorities<T> {
...
 
 impl<X, T> session::OnSessionChange<X> for SyncedAuthorities<T> where
-	T: Trait,
-	T: session::Trait,
-	<T as session::Trait>::ConvertAccountIdToSessionKey: Convert<
-		<T as system::Trait>::AccountId,
-		<T as Trait>::SessionKey,
-	>,
+	T: Trait + consensus::Trait<SessionKey=<T as Trait>::SessionKey>,
+	<T as consensus::Trait>::Log: From<consensus::RawLog<<T as Trait>::SessionKey>>
```
Comment (rphmeier) [inline srml/grandpa/src/lib.rs]: 👍

**Comment by Kianenigma** [comment]
Seem to have gotten sufficient remark and review already; I will scan these changes for portable portions to the js library. `DeriveJunction` and `HDKD` are the main features to be ported as I have been informed so far.

File: core/primitives/src/crypto.rs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+			s
+		};
+
+		if let Ok(d) = hex::decode(hex_seed) {
+			if let Some(r) = Self::from_seed_slice(&d) {
+				return Some(r)
+			}
+		}
+
+		let re = Regex::new(r"^(?P<phrase>\w+( \w+)*)(?P<path>(//?[^/]+)*)(///(?P<password>.*))?$")
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply core/primitives/src/crypto.rs]: passwords may legally be empty.

File: core/primitives/src/sr25519.rs [hunk 5]
```diff
@@ -537,4 +500,16 @@ mod test {
 		let enc = hex!["090fa15cb5b1666222fff584b4cc2b1761fe1e238346b340491b37e25ea183ff"];
 		assert_eq!(Public::from_ss58check(k).unwrap(), Public::from_raw(enc));
 	}
+
+	#[ignore]
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply core/primitives/src/sr25519.rs]: removed

File: core/primitives/src/sr25519.rs [hunk 6]
```diff
@@ -16,33 +16,202 @@
...
-use bip39::{Mnemonic, Language};
 
 #[cfg(feature = "std")]
-use serde::{de, Deserialize, Deserializer, Serializer};
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+#[cfg(feature = "std")]
+use schnorrkel::keys::MINI_SECRET_KEY_LENGTH;
 
 // signing context
-const SIGNING_CTX: &'static [u8] = b"substrate transaction";
```
Comment (gavofyork) [inline core/primitives/src/sr25519.rs]: @Kianenigma I changed this, so JS will need update.

**Comment by gavofyork** [comment]
@tomusdrw I addressed the grumbles. @Kianenigma If you're also happy with the code, then I'd like to merge today.

**Pull request merged**